src/exo5 - Nested data with traverse/exo5.effect.solution.ts

```diff
@@ -0,0 +1,255 @@
+type CountryCode = 'FR' | 'SP' | 'US';
+export const countryNameToCountryCode: Record<string, CountryCode> = {
+  France: 'FR',
+  Spain: 'SP',
+  USA: 'US',
+};
+
+// Let's simulate the call to an api which would return the currency when
+// providing a country code. For the sake of simplicity, let's consider that it
+// cannot fail.
+type Currency = 'EUR' | 'DOLLAR';
+export const getCountryCurrency = (countryCode: CountryCode): Effect.Effect<Currency> =>
+  Effect.gen(function* () {
+    if (countryCode === 'US') {
+      return 'DOLLAR';
+    }
+    return 'EUR';
+  });
+
+// Let's simulate a way for the user to provide a country name.
+// Let's consider that it cannot fail and let's add the possibility to set
+// the user's response as a parameter for easier testing.
+export const getCountryNameFromUser = (countryName: string): Effect.Effect<string> =>
+  Effect.succeed(countryName);
+
+// Here's a function to retrieve the countryCode from a country name if it is
+// matching a country we support. This method returns an `Option` as we cannot
+// return anything if the given string is not matching a country name we know
+export const getCountryCode = (countryName: string): Option.Option<CountryCode> =>
+  Option.fromNullable(countryNameToCountryCode[countryName]);
+
+///////////////////////////////////////////////////////////////////////////////
+//                            TRAVERSING OPTIONS                             //
+///////////////////////////////////////////////////////////////////////////////
+
+// With all these functions, we can simulate a program that would ask for a
+// country name and return its currency if it knows the country.
+// A naive implementation would be mapping on each `Effect` and `Option` to call
+// the correct method:
+export const naiveGiveCurrencyOfCountryToUser = (countryNameFromUserMock: string) =>
+  pipe(
+    getCountryNameFromUser(countryNameFromUserMock),
+    Effect.map(getCountryCode),
+    Effect.map(Option.map(getCountryCurrency))
+  );
+
+// The result type of this method is: `Effect<Option<Effect<Currency>>>`
+// Not ideal, right? We would need to run the first `Effect`, then check if it's
+// `Some` to get the `Effect` inside and finally run the `Effect` to retrieve the
+// currency.
+// Let's do better than that!
+
+// First we need a way to transform our `Option<Effect<Currency>>` to
+// `Effect<Option<Currency>>`
+// That's precisely what `Effect.flatMapOption` is about.
+// Use `Option.traverse` to implement `getCountryCurrencyOfOptionalCountryCode`
+// below. This function takes an `Option<CountryCode>`, should apply
+// `getCountryCurrency` to the `CountryCode` and make it so that the result
+// is `Effect<Option<Currency>>`
+
+export const getCountryCurrencyOfOptionalCountryCode = (
+  optionalCountryCode: Option.Option<CountryCode>
+): Effect.Effect<Option.Option<Currency>> =>
+  // Implement this function using Effect's APIs
+  pipe(
+    optionalCountryCode,
+    Option.match({
```

Comment:
Effect equivalent of `option.traverse` here would be `Effect.transposeMapOption` that I added recently in Effect 😏 let's use it! :)